### ISSUE TITLE
Custom Emebeddings with Seed Topics

### ISSUE THREAD

**Issue by VomV** [MaartenGr/BERTopic#1208]
@MaartenGr First of all, many thanks for developing BERTopic and open sourcing it. It is really amazing and has helped me so much.

I am using custom embeddings for my BERTopic model by creating a backend to generate the embeddings and input to the BERTopic model with fit_transform. This however doesn't seem to work with seed topics. I currently don't see a way to feed custom seed topic embeddings to the BERTopic class or via fit_transform. Do you have any suggestions for this? If its not possible currently, I would be happy to work on adding the capability to input custom seed topic to the model.

**Comment by VomV** [comment]
I have created an API that generates the embeddings on a GPU. In my code, I call the API to get the embeddings and then run the rest of the steps on a CPU. So, I don't declare an embedding model when I instantiate the Bertopic class but I did provide a seed topic list. Then I call fit transform on the model object with docs & embeddings. Since I am not providing an embedding model to Bertopic class, it is not able to generate seed topic embeddings I think. Ideally, I would like to create seed topic embeddings from my API and feed it to the model somehow. But I don't see a way of doing that.

This is what I am trying to do (full notebook: [here](https://colab.research.google.com/drive/1VOe4M5kFvt0kEJjvqQlGWnNFJvMpR4kX?usp=sharing)):

    # Embedding Model
    docs = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))['data']
    sentence_model = SentenceTransformer("all-MiniLM-L6-v2")

    # Prepare embeddings
    embeddings = sentence_model.encode(docs, show_progress_bar=False)
    
    #Seed Topic
    seed_topic_list = [["drug", "cancer", "drugs", "doctor"],
                       ["windows", "drive", "dos", "file"],
                       ["space", "launch", "orbit", "lunar"]]
    
    seed_topic_texts = [' '.join(seed_topic) for seed_topic in seed_topic_list]
    seed_topic_embeddings = sentence_model.encode(seed_topic_texts, show_progress_bar=True)
    
    # Train our topic model using our pre-trained sentence-transformers embeddings
    topic_model = BERTopic(seed_topic_list=seed_topic_list, calculate_probabilities=False)
    topics, _ = topic_model.fit_transform(docs, embeddings)
    
    ######What I would like to do::::
     topics, _ = topic_model.fit_transform(docs, embeddings, seed_topic_embeddings)

**Comment by MaartenGr** [comment]
Ah right, I thought that you did declare an embedding model, my bad. 

The thing is, the `seed_topic_list` will nearly always be a very small list of that even on a CPU will get embedded quite quickly. So I am not sure whether there is a need to calculate the seed_topic_embeddings beforehand if they are small.

**Comment by VomV** [comment]
Yeah, agree there. Just that I am trying to avoid loading the embedding model in my front end application to save memory and so need to build a custom functionality probably. If you think it would be helpful to add this to bertopic (maybe through fit_transform), please let me know and I will create an issue for that and add it.

**Comment by MaartenGr** [comment]
Hmmm, this is tricky. I want to prevent overcomplicating the API as much as possible but I can imagine this to be quite a troublesome issue for you. 

There is something you can do, albeit a bit hacky. The following code is where the seeded topic embeddings are generated:

https://github.com/MaartenGr/BERTopic/blob/f1e80797fd4361b46c54d041e76f6a3752b617d9/bertopic/_bertopic.py#L2968

What you could do is create a [custom backend](https://maartengr.github.io/BERTopic/getting_started/embeddings/embeddings.html#custom-backend) that simply returns your pre-calculated seeded embeddings. Seeing as `._extract_embeddings` generates embeddings from the `embedding_model`, you could create an embedding model that only returns `seed_topic_embeddings`. 

So, psuedo-code-ish, something like this:

```python
from bertopic.backend import BaseEmbedder

class CustomEmbedder(BaseEmbedder):
    def __init__(self, seed_topic_embeddings):
        super().__init__()
        self.seed_topic_embeddings = seed_topic_embeddings 

    def embed(self, documents, verbose=False):
        return self.seed_topic_embeddings 

# Embedding Model
docs = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))['data']
sentence_model = SentenceTransformer("all-MiniLM-L6-v2")

# Prepare embeddings
embeddings = sentence_model.encode(docs, show_progress_bar=False)

#Seed Topic
seed_topic_list = [["drug", "cancer", "drugs", "doctor"],
                   ["windows", "drive", "dos", "file"],
                   ["space", "launch", "orbit", "lunar"]]

seed_topic_texts = [' '.join(seed_topic) for seed_topic in seed_topic_list]
seed_topic_embeddings = sentence_model.encode(seed_topic_texts, show_progress_bar=True)

# Train our topic model using our pre-trained sentence-transformers embeddings
custom_embedder = CustomEmbedder(seed_topic_embeddings=seed_topic_embeddings)
topic_model = BERTopic(seed_topic_list=seed_topic_list, embedding_model=custom_embedder)
topics, _ = topic_model.fit_transform(docs, embeddings)

**Comment by VomV** [comment]
Amazing, I will give this a try. Thanks @MaartenGr

**Comment by MaartenGr** [comment]
Closing this due to inactivity. Let me know if I need to re-open the issue!
